File: Perceptron/Perceptron.m
clc;
close all;
clear all;
%% 定义训练样本
Class_1_o=  [220 90;240 95;220 95;180 95;140 90];
Class_2_o=  [80 85;85 80;85 85;82 80;78 80];
Test_Data_o = [180 90;210 90;140 90;90 80;78 80];

Class_x = mapminmax([Class_1_o(:,1)',Class_2_o(:,1)',Test_Data_o(:,1)'], 0, 1); 
Class_y = mapminmax([Class_1_o(:,2)',Class_2_o(:,2)',Test_Data_o(:,2)'], 0, 1);
Class_1 = [Class_x(1:5)',Class_y(1:5)'];
Class_2 = [Class_x(6:10)',Class_y(6:10)'];
Test_Data = [Class_x(11:15)',Class_y(11:15)'];

[m_1,n_1] = size(Class_1);
[m_2,n_2] = size(Class_2);
%% 绘图程序
figure;
plot(Class_1(:,1),Class_1(:,2),'r*','LineWidth',2); hold on;
plot(Class_2(:,1),Class_2(:,2),'bo','LineWidth',2); hold on;
plot(Test_Data(:,1),Test_Data(:,2),'gs','LineWidth',2); hold on;
%% 增广向量
att_1=ones(m_1,1);
att_2=ones(m_2,1);
Train_Data = [Class_1,att_1; -1*Class_2,-1*att_2];
Test_Data = [Test_Data,ones(5,1)];
[m_3,n_3] = size(Train_Data);
w = randn([3 1]);  % 随机初始化权重以及参数
a = 0.1;  % 学习率
epoch = 0;   % 定义迭代次数
%% 开始训练
while 1==1
    % 找到分类错误的样本
    b=w;
     for i=1:m_3
     if w'*Train_Data(i,:)' <= 0
         w = w + a * Train_Data(i,:)';
     end
    end
    if b==w
        break;
    end
    epoch = epoch + 1;
end
%% 绘制最终分类线
f = [num2str(w(1,1)) '*x1+(' num2str(w(2,1)) ')*x2+(' num2str(w(3,1)) ')'] ; 
h = ezplot(f,[0,1]);
grid on;
set(h,'Color','r');
legend('Class1','Class2','Test','DecisionBoundary')
set(legend,'location','SouthEast')
%% 打印结果
for k=1:5
    Result = w'*Test_Data(k,:)';
    if Result > 0
        fprintf('测试样本%d结果为%f属于第一类 \n',k,Result);
    elseif Result <= 0
        fprintf('测试样本%d结果为%f属于第二类 \n',k,Result);
    end
end

File: Perceptron/Perceptron.py
import matplotlib.pyplot as plt
import torch
import numpy as np


# 进行归一化
def normalization(Data1, Data2, Data3):
    # 要对全部的X坐标单独归一化，全部的Y坐标同理
    Data_x = np.concatenate((Data1[:, 0], Data2[:, 0], Data3[:, 0]))
    Data_y = np.concatenate((Data1[:, 1], Data2[:, 1], Data3[:, 1]))
    # 参照Matlab的mapminmax函数进行归一化
    Data_x = (Data_x - np.min(Data_x)) / (np.max(Data_x) - np.min(Data_x))
    Data_y = (Data_y - np.min(Data_y)) / (np.max(Data_y) - np.min(Data_y))
    Data_x = np.expand_dims(Data_x, 1)
    Data_y = np.expand_dims(Data_y, 1)

    Data1 = np.concatenate((Data_x[:5], Data_y[:5]), 1)
    Data2 = np.concatenate((Data_x[5:10], Data_y[5:10]), 1)
    Data3 = np.concatenate((Data_x[10:], Data_y[10:]), 1)
    return Data1, Data2, Data3


if __name__ == "__main__":
    # 定义数据集
    class_1_o = np.array([[220, 90], [240, 95], [220, 95], [180, 95], [140, 90]], dtype=np.float)
    class_2_o = np.array([[80, 85], [85, 80], [85, 85], [82, 80], [78, 80]], dtype=np.float)
    Test_Data_o = np.array([[180, 90], [210, 90], [140, 90], [90, 80], [78, 80]], dtype=np.float)
    Class_1, Class_2, Test_Data = normalization(class_1_o, class_2_o, Test_Data_o)
    a = Class_1.shape
    b = Class_2.shape
    # 开始绘图
    plt.figure()
    plt.plot(Class_1[:, 0], Class_1[:, 1], 'r*', label='Class_1')
    plt.plot(Class_2[:, 0], Class_2[:, 1], 'b*', label='Class_2')
    plt.plot(Test_Data[:, 0], Test_Data[:, 1], 'gs', label='Test_Data')
    plt.legend(loc='best')
    # 增广向量
    att_1 = np.ones((a[0], 1))
    att_2 = np.ones((b[0], 1))
    # 这里对数据集做了预处理使得 当正确分类是w * x > 0   ，错误分类是 w * x <= =
    Train_Data = np.concatenate((np.concatenate((Class_1, att_1),  axis=1), -1 * np.concatenate((Class_2, att_2),  axis=1) ), axis=0)
    Test_Data = np.concatenate((Test_Data, np.ones((5, 1))), axis=1)
    c = Train_Data.shape
    # w = np.random.randn(3, 1)
    w = np.array([[1], [1], [1]])  # 使用[1;1;1]方便与C++的运行结果比对（因为C++无法使用matlpotlib绘图）
    a = 0.1
    # 开始训练
    while True:
        bb = w
        for i in range(0, c[0]):
            if np.matmul(w.T, Train_Data[i, :].T) <= 0:  # matmul为矩阵乘法（线代知识）
                w = w + a * Train_Data[i, :].reshape(3, 1)
        if (bb == w).all():
            break
    print(w)
    # 绘制分类线
    xlist = np.linspace(-0.5, 1.5, 100)
    ylist = np.linspace(-0.5, 1.5, 100)
    x, y = np.meshgrid(xlist, ylist)  # 计算圆所在区域的网格
    f = w[0] * x + w[1] * y + w[2] < 0
    plt.contourf(x, y, f, cmap="cool")
    plt.show()


